proto/proto/dex.proto

```diff
@@ -12,28 +12,66 @@ message Swap {
   MockFlowCiphertext enc_amount_1 = 2;
   // Encrypted amount of asset 2 of the trading pair.
   MockFlowCiphertext enc_amount_2 = 3;
+  // Encapsulates the authorized fields of the Swap transaction, used in signing.
+  SwapBody body = 4;
+  // The spend authorization signature is effecting data.
+  crypto.SpendAuthSignature auth_sig = 5;
+}
+
+// A transaction action that obtains assets previously confirmed
+// via a Swap transaction. Does not include a spend authorization
+// signature, as it is only capable of consuming the NFT from a
+// Swap transaction.
+message SwapClaim {
+  // Proves the spent note records 1 unit of the swap NFT.
+  bytes zkproof = 1;
+  // The nullifier for the Swap NFT to be consumed.
+  crypto.Nullifier nullifier = 2;
+  // The fee allows `SwapClaim` without an additional `Spend`.
+  transaction.Fee fee = 3;
+  // Note output for asset 1.
+  crypto.NotePayload output_1 = 4;
+  // Note output for asset 2.
+  crypto.NotePayload output_2 = 5;
+  // Root for the associated `Swap` action.
+  crypto.MerkleRoot anchor = 6;
+  // Price for asset 1 in the associated `Swap`.
+  uint64 price_1 = 7;
+  // Price for asset 2 in the associated `Swap`.
+  uint64 price_2 = 8;
 }
 
+// The authorized data of a Swap transaction.
 message SwapBody {
   // The trading pair to swap.
   TradingPair trading_pair = 1;
   // Commitment to the amount for asset 1 (delta 1).
   bytes ca1 = 2;
   // Commitment to the amount for asset 2 (delta 2).
   bytes ca2 = 3;
-  // A commitment to the swap's contribution to the transaction's value balance.
-  bytes cv = 4;
+  // A commitment to the fee's contribution to the transaction's value balance.
+  bytes cf = 4;
   // Swap NFT recording the user's contribution.
   crypto.NotePayload swap_nft = 5;
+  // Encrypted version of the original `Swap`, symmetrically encrypted w/ viewing key.
+  bytes swap_ciphertext = 6;
 }
 
 // XXX value fields need to have constant-length encoding
+// ? or probably not, since this is a plaintext
```

Comment:
Yep, we can bundle the constant-length encoding into the encrypt/decrypt functions, since we'll only need to handle it there.